crates/typst-cli/src/package.rs

```diff
@@ -56,10 +59,84 @@ impl PackageStorage {
             if dir.exists() {
                 return Ok(dir);
             }
+            let parent_dir = dir.parent().unwrap();
+
+            let file_locking_err = |string| {
+                move |err| PackageError::FileLocking(eco_format!("{string}: {err}"))
+            };
+
+            let lock_file_path = parent_dir.join(format!(".{deepest_dir}.lock"));
+            let remove_lock = || {
+                fs::remove_file(&lock_file_path)
+                    .map_err(file_locking_err("failed to remove lock file"))
+            };
+            if !parent_dir.exists() {
+                fs::create_dir_all(parent_dir)
+                    .map_err(file_locking_err("failed to create parent directories"))?;
+            }
+            // https://github.com/yoshuawuyts/fd-lock/issues/28#issuecomment-2264180893
+            // File::create(&lock_file_path).unwrap();
+            // let file = match File::open(lock_file_path.as_path()) {
+            let file = fs::OpenOptions::new()
+                .create(true)
+                .truncate(true)
+                .write(true)
+                // .read(true) // to test the error
+                .open(lock_file_path.as_path())
+                .map_err(file_locking_err("failed to open lock file"))?;
+            let lock_file_path = format!(
+                "~{}",
+                lock_file_path.to_string_lossy().strip_prefix(env!("HOME")).unwrap()
+            );
+            // let mut lock_file = RwLock::new(file.try_clone().unwrap());
+            let mut lock_file = RwLock::new(file);
```

Comment:
I'd prefer to use the full namespace path here, as this had initially confused me (missed the import and thought this was `std::sync::RwLock`).

```suggestion
            let mut lock_file = fd_lock::RwLock::new(file);
```